src/auspex/experiment.py

```diff
@@ -491,44 +502,175 @@ def catch_ctrl_c(signum, frame):
+            mem_plots    = {n.filter_name: mem_fig.line(x='time', y='mem', color=colors[i], line_width=4, source=data_sources[n.filter_name]) for i, n in enumerate(self.other_nodes)}
+            vmem_plots   = {n.filter_name: vmem_fig.line(x='time', y='proc', color=colors[i], line_width=4, source=data_sources[n.filter_name]) for i, n in enumerate(self.other_nodes)}
+
+            legend_1 = Legend(items=[(n , [l]) for n, l in mem_plots.items()], location=(0, 0))
+            legend_2 = Legend(items=[(n , [l]) for n, l in vmem_plots.items()], location=(0, 0))
+            legend_3 = Legend(items=[(n , [l]) for n, l in cpu_plots.items()], location=(0, 0))
+            mem_fig.add_layout(legend_1, 'right')
+            vmem_fig.add_layout(legend_2, 'right')
+            cpu_fig.add_layout(legend_3, 'right')
+            container = column([cpu_fig, mem_fig, vmem_fig])
+            doc = curdoc()
+
+            exit_perf = Event()
+
+            @gen.coroutine
+            def update_perf(name, time_val, cpu, mem, vmem, proc):
+                data_sources[name].stream(dict(time=[time_val], cpu=[cpu], mem=[mem], vmem=[vmem], proc=[proc]))
+
+            def wait_for_perf_updates(q, exit):
+                while not exit.is_set():
+                    messages = []
+
+                    while not exit.is_set():
+                        try:
+                            messages.append(q.get(False))
+                        except queue.Empty as e:
+                            time.sleep(0.05)
+                            break
+
+                    for message in messages:
+                        filter_name, time_val, cpu, mem_info, processed = message
+                        mem = mem_info[0]/2.**20
+                        vmem = mem_info[1]/2.**20
+                        proc = processed/2.**20
+                        update_perf(filter_name, time_val, cpu, mem, vmem, proc)
+
+            perf_thread = Thread(target=wait_for_perf_updates, args=(perf_queue, exit_perf))
+            perf_thread.start()
+
+            for f in self.other_nodes:
+                f.perf_queue = perf_queue
+
+            doc.clear()
+            doc.add_root(container)
+            session = push_session(doc)
+            session.show(container)
+
+        # Start the filter processes
+        for n in self.other_nodes:
+            n.start()
+
+        # Run the main experiment loop
+        self.sweep()
+
+        # Make sure any last minute plotting needs are met
         for plot, callback in zip(self.manual_plotters, self.manual_plotter_callbacks):
             if callback:
                 callback(plot)
 
+        # Wait for the
+        time.sleep(0.1)
+        times = {n: 0 for n in self.other_nodes}
+        dones = {n: False for n in self.other_nodes}
+        while False in dones.values():
+            time.sleep(1)
+            for n in self.other_nodes:
+                if not n.done.is_set():
+                    times[n] += 1
```

Comment:
It may be good to set a maximum number of `times[n]` until it stops.

Reply:
Done!